File: scala/src/main/scala/benchmarks/Distance.scala
package benchmarks

import scala.collection.mutable

object Distance:
  case class StringPair(s1: String, s2: String)

  def generatePairStrings(n: Int, m: Int): Array[StringPair] =
    val chars = "abcdefghij".toCharArray
    val pairs = new Array[StringPair](n)

    for (i <- 0 until n) {
      val len1 = Helper.nextInt(m) + 4
      val len2 = Helper.nextInt(m) + 4

      val str1 = new StringBuilder(len1)
      val str2 = new StringBuilder(len2)

      for (_ <- 0 until len1) str1.append(chars(Helper.nextInt(10)))
      for (_ <- 0 until len2) str2.append(chars(Helper.nextInt(10)))

      pairs(i) = StringPair(str1.toString, str2.toString)
    }

    pairs

  class Jaro extends Benchmark:
    private var count: Int = 0
    private var size: Int = 0
    private var pairs: Array[StringPair] = _
    private var resultVal: Long = 0L

    override def name(): String = "Distance::Jaro"

    override def prepare(): Unit =
      count = configVal("count").toInt
      size = configVal("size").toInt
      pairs = Distance.generatePairStrings(count, size)
      resultVal = 0L

    private def jaro(s1: String, s2: String): Double =
      val bytes1 = s1.getBytes("US-ASCII")
      val bytes2 = s2.getBytes("US-ASCII")

      val len1 = bytes1.length
      val len2 = bytes2.length

      if len1 == 0 || len2 == 0 then return 0.0

      var matchDist = Math.max(len1, len2) / 2 - 1
      if matchDist < 0 then matchDist = 0

      val s1Matches = new Array[Boolean](len1)
      val s2Matches = new Array[Boolean](len2)

      var matches = 0
      for (i <- 0 until len1) {
        val start = Math.max(0, i - matchDist)
        val end = Math.min(len2 - 1, i + matchDist)

        var found = false
        var j = start
        while j <= end && !found do
          if (!s2Matches(j) && bytes1(i) == bytes2(j)) {
            s1Matches(i) = true
            s2Matches(j) = true
            matches += 1
            found = true
          }
          j += 1
      }

      if matches == 0 then return 0.0

      var transpositions = 0
      var k = 0
      for (i <- 0 until len1) {
        if (s1Matches(i)) {
          while (k < len2 && !s2Matches(k)) {
            k += 1
          }
          if (k < len2) {
            if (bytes1(i) != bytes2(k)) {
              transpositions += 1
            }
            k += 1
          }
        }
      }
      transpositions /= 2

      val m = matches.toDouble
      (m / len1 + m / len2 + (m - transpositions) / m) / 3.0

    override def run(iterationId: Int): Unit =
      for (pair <- pairs) {
        resultVal = (resultVal + (jaro(pair.s1, pair.s2) * 1000).toLong) & 0xffffffffL
      }

    override def checksum(): Long = resultVal & 0xffffffffL

  class NGram extends Benchmark:
    private var count: Int = 0
    private var size: Int = 0
    private var pairs: Array[StringPair] = _
    private var resultVal: Long = 0L
    private val N = 4

    override def name(): String = "Distance::NGram"

    override def prepare(): Unit =
      count = configVal("count").toInt
      size = configVal("size").toInt
      pairs = Distance.generatePairStrings(count, size)
      resultVal = 0L

    private def ngram(s1: String, s2: String): Double =
      val bytes1 = s1.getBytes("US-ASCII")
      val bytes2 = s2.getBytes("US-ASCII")

      if bytes1.length < N || bytes2.length < N then return 0.0

      val grams1 = new java.util.HashMap[Int, Int](bytes1.length)

      for (i <- 0 to bytes1.length - N) {
        val gram = (bytes1(i) & 0xff) << 24 |
          (bytes1(i + 1) & 0xff) << 16 |
          (bytes1(i + 2) & 0xff) << 8 |
          (bytes1(i + 3) & 0xff)

        grams1.merge(gram, 1, (a, b) => a + b)
      }

      val grams2 = new java.util.HashMap[Int, Int](bytes2.length)
      var intersection = 0

      for (i <- 0 to bytes2.length - N) {
        val gram = (bytes2(i) & 0xff) << 24 |
          (bytes2(i + 1) & 0xff) << 16 |
          (bytes2(i + 2) & 0xff) << 8 |
          (bytes2(i + 3) & 0xff)

        grams2.merge(gram, 1, (a, b) => a + b)

        if grams1.containsKey(gram) then
          val cnt1 = grams1.get(gram)
          if grams2.get(gram) <= cnt1 then intersection += 1
      }

      val total = grams1.size + grams2.size
      if total > 0 then intersection.toDouble / total else 0.0

    override def run(iterationId: Int): Unit =
      for (pair <- pairs) {
        resultVal = (resultVal + (ngram(pair.s1, pair.s2) * 1000).toLong) & 0xffffffffL
      }

    override def checksum(): Long = resultVal & 0xffffffffL

File: java/src/main/java/benchmarks/Distance.java
package benchmarks;

import java.util.*;
import java.nio.charset.StandardCharsets;

public class Distance {

    public static class Pair {
        public final String s1;
        public final String s2;

        public Pair(String s1, String s2) {
            this.s1 = s1;
            this.s2 = s2;
        }
    }

    public static Pair[] generatePairStrings(long n, long m) {
        Pair[] pairs = new Pair[(int) n];
        char[] chars = "abcdefghij".toCharArray();

        for (int i = 0; i < n; i++) {
            int len1 = Helper.nextInt((int) m) + 4;
            int len2 = Helper.nextInt((int) m) + 4;

            StringBuilder sb1 = new StringBuilder(len1);
            StringBuilder sb2 = new StringBuilder(len2);

            for (int j = 0; j < len1; j++) {
                sb1.append(chars[Helper.nextInt(10)]);
            }
            for (int j = 0; j < len2; j++) {
                sb2.append(chars[Helper.nextInt(10)]);
            }

            pairs[i] = new Pair(sb1.toString(), sb2.toString());
        }

        return pairs;
    }

    public static class Jaro extends Benchmark {
        private long count;
        private long size;
        private Pair[] pairs;
        private long resultVal;

        public Jaro() {
            this.resultVal = 0L;
        }

        @Override
        public String name() {
            return "Distance::Jaro";
        }

        @Override
        public void prepare() {
            this.count = configVal("count");
            this.size = configVal("size");
            this.pairs = generatePairStrings(count, size);
            this.resultVal = 0L;
        }

        private double jaro(String s1, String s2) {
            byte[] bytes1 = s1.getBytes(StandardCharsets.US_ASCII);
            byte[] bytes2 = s2.getBytes(StandardCharsets.US_ASCII);

            int len1 = bytes1.length;
            int len2 = bytes2.length;

            if (len1 == 0 || len2 == 0) return 0.0;

            int matchDist = Math.max(len1, len2) / 2 - 1;
            if (matchDist < 0) matchDist = 0;

            boolean[] s1Matches = new boolean[len1];
            boolean[] s2Matches = new boolean[len2];

            int matches = 0;
            for (int i = 0; i < len1; i++) {
                int start = Math.max(0, i - matchDist);
                int end = Math.min(len2 - 1, i + matchDist);

                for (int j = start; j <= end; j++) {
                    if (!s2Matches[j] && bytes1[i] == bytes2[j]) {
                        s1Matches[i] = true;
                        s2Matches[j] = true;
                        matches++;
                        break;
                    }
                }
            }

            if (matches == 0) return 0.0;

            int transpositions = 0;
            int k = 0;
            for (int i = 0; i < len1; i++) {
                if (s1Matches[i]) {
                    while (k < len2 && !s2Matches[k]) {
                        k++;
                    }
                    if (k < len2) {
                        if (bytes1[i] != bytes2[k]) {
                            transpositions++;
                        }
                        k++;
                    }
                }
            }
            transpositions /= 2;

            double m = matches;
            return (m / len1 + m / len2 + (m - transpositions) / m) / 3.0;
        }

        @Override
        public void run(int iterationId) {
            for (Pair pair : pairs) {
                resultVal += (long) (jaro(pair.s1, pair.s2) * 1000);
            }
        }

        @Override
        public long checksum() {
            return resultVal;
        }
    }

    public static class NGram extends Benchmark {
        private long count;
        private long size;
        private Pair[] pairs;
        private long resultVal;
        private static final int N = 4;

        public NGram() {
            this.resultVal = 0L;
        }

        @Override
        public String name() {
            return "Distance::NGram";
        }

        @Override
        public void prepare() {
            this.count = configVal("count");
            this.size = configVal("size");
            this.pairs = generatePairStrings(count, size);
            this.resultVal = 0L;
        }

        private double ngram(String s1, String s2) {
            if (s1.length() < N || s2.length() < N) return 0.0;

            byte[] bytes1 = s1.getBytes(StandardCharsets.US_ASCII);
            byte[] bytes2 = s2.getBytes(StandardCharsets.US_ASCII);

            Map<Integer, Integer> grams1 = new HashMap<>(bytes1.length);

            for (int i = 0; i <= bytes1.length - N; i++) {
                int gram = ((int) bytes1[i] & 0xFF) << 24 |
                           ((int) bytes1[i + 1] & 0xFF) << 16 |
                           ((int) bytes1[i + 2] & 0xFF) << 8 |
                           ((int) bytes1[i + 3] & 0xFF);

                grams1.merge(gram, 1, Integer::sum);
            }

            Map<Integer, Integer> grams2 = new HashMap<>(bytes2.length);
            int intersection = 0;

            for (int i = 0; i <= bytes2.length - N; i++) {
                int gram = ((int) bytes2[i] & 0xFF) << 24 |
                           ((int) bytes2[i + 1] & 0xFF) << 16 |
                           ((int) bytes2[i + 2] & 0xFF) << 8 |
                           ((int) bytes2[i + 3] & 0xFF);

                grams2.merge(gram, 1, Integer::sum);

                Integer cnt1 = grams1.get(gram);
                if (cnt1 != null && grams2.get(gram) <= cnt1) {
                    intersection++;
                }
            }

            int total = grams1.size() + grams2.size();
            return total > 0 ? (double) intersection / total : 0.0;
        }

        @Override
        public void run(int iterationId) {
            for (Pair pair : pairs) {
                resultVal += (long) (ngram(pair.s1, pair.s2) * 1000);
            }
        }

        @Override
        public long checksum() {
            return resultVal;
        }
    }
}